source/agora/test/ValidatorCleanRestart.d

```diff
@@ -0,0 +1,141 @@
+    auto txs = spendable[0 .. 5]
+        .map!(txb => txb.refund(WK.Keys.Genesis.address).sign())
+        .array;
+
+    // 8 utxos for freezing, 8 utxos for creating a block later
+    txs ~= spendable[5].split(WK.Keys.byRange.take(8).map!(k => k.address)).sign();
+    txs ~= spendable[6].split(WK.Keys.Genesis.address.repeat(8)).sign();
+    txs ~= spendable[7].split(WK.Keys.Genesis.address.repeat(8)).sign();
+
+    // Block 8
+    txs.each!(tx => nodes[0].putTransaction(tx));
+    network.expectBlock(Height(8), 5.seconds);
+
+    // Freeze builders
+    auto freezable = txs[$ - 3]
+        .outputs.length.iota
+        .takeExactly(8)
+        .map!(idx => TxBuilder(txs[$ - 3], cast(uint)idx))
+        .array;
+
+    // Create 8 freeze TXs
+    auto freeze_txs = freezable
+        .enumerate
+        .map!(pair => pair.value.refund(WK.Keys[pair.index].address)
+            .sign(TxType.Freeze))
+        .array;
+    assert(freeze_txs.length == 8);
+
+    // Block 9
+    freeze_txs.each!(tx => nodes[0].putTransaction(tx));
+    network.expectBlock(Height(9), 5.seconds);
+
+    // Now we enroll 4 new validators. After this, the already enrolled
+    // validators will be expired.
+    foreach (i; 4 .. nodes.length)
+    {
+        Enrollment enroll = nodes[i].createEnrollmentData();
+        nodes[i].enrollValidator(enroll);
+
+        // Check enrollment
+        nodes.each!(n =>
+            retryFor(n.getEnrollment(enroll.utxo_key) == enroll, 5.seconds));
+    }
+
+    // Block 10
+    auto new_txs = txs[$ - 2]
+        .outputs.length.iota.map!(idx => TxBuilder(txs[$ - 2], cast(uint)idx))
+        .takeExactly(8)
+        .map!(txb => txb.refund(WK.Keys.Genesis.address).sign()).array;
+    new_txs.each!(tx => nodes[0].putTransaction(tx));
+    network.expectBlock(Height(10), 5.seconds);
+
+    // Sanity check
+    nodes.enumerate.each!((idx, node) =>
+        retryFor(node.getValidatorCount == 4, 3.seconds,
+            format("Node %s has validator count %s. Expected: %s",
+                idx, node.getValidatorCount(), 4)));
+
+    // Now restarting the validators in the set B, all the date of those
+    // validators has been wiped out.
+    foreach (i; 4 .. nodes.length)
+        network.restart(nodes[i]);
+
+    network.expectBlock(Height(10), 8.seconds);
+
+    // Check the connection state is complete for all the nodes
+    nodes.each!(node =>
+        retryFor(node.getNodeInfo().state == NetworkState.Complete, 1.seconds));
+
+    // Shutdown all the validators in the set A
+    foreach (i; 0 .. 4)
+        nodes[i].shutdown();
+
+    // Block 11 with the new validators in the set B
+    new_txs = txs[$ - 1]
+        .outputs.length.iota.map!(idx => TxBuilder(txs[$ - 1], cast(uint)idx))
+        .takeExactly(8)
+        .map!(txb => txb.refund(WK.Keys.Genesis.address).sign()).array;
+    new_txs.each!(tx => nodes[0].putTransaction(tx));
+    network.expectBlock(Height(11), 8.seconds);
```

Comment:
```suggestion
    network.expectBlock(nodes[4 .. $], Height(11), 8.seconds);
```

That is probably the reason why you saw "MessageBox closed" errors. Because the main thread was sending messages to node threads which were shut down.